pp.c

```diff
@@ -7145,10 +7145,15 @@ PP(pp_argcheck)
     too_few = (argc < (params - opt_params));
 
     if (UNLIKELY(too_few || (!slurpy && argc > params)))
+
         /* diag_listed_as: Too few arguments for subroutine '%s' */
         /* diag_listed_as: Too many arguments for subroutine '%s' */
-        Perl_croak_caller("Too %s arguments for subroutine '%" SVf "'",
-                          too_few ? "few" : "many", S_find_runcv_name());
+        Perl_croak_caller("Too %s arguments (got %" UVuf ", expected %s%" UVuf ") for subroutine '%" SVf "'",
```

Comment:
This should probably include an update to pod/perldiag.pod for the changed diagnostic messages (and then a corresponding update to the `diag_listed_as` cookies above).

Maybe it would also be better to move the argument count to the end of the message, as in `Too few arguments for subroutine 'Foo::foo' (got 1, expected at least 2)`.

Reply:
I put the counts after “arguments” so that the “modifier” (i.e., the counts) stayed closer to the referent (“arguments”).

What is the advantage of putting the counts at the end? (I can surmise but want to know your reasoning.)

Reply:
Well, I first considered it on the basis that it might make the perldiag cookie easier to construct and read. But having thought of it, it seemed to me that it would tend to make the messages read better, by having the big chunky part at the end. I don't have a terribly strong opinion on that part, though.

Reply:
I would also agree with arc, but that's also a not terribly strong opinion on my side.